### PR TITLE
Optimise hash joins

### PR DESCRIPTION
**Pull request by systay**
## Description
We recently added support for more queries by using hash joins when a nested loop join is not possible to use.

At the moment these queries would work, but only by moving the data to the vtgate proxy and there doing the necessary operations on the data. This makes a lot of queries intractable - they work only on toy datasets and would time out on real data sizes.

This PR adds logic to push most of the work down to mysql.

## Related Issue(s)
PR that introduced hash joins: #14448

## Checklist
-   [x] "Backport to:" labels have been added if this change should be back-ported
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on the CI
-   [x] Documentation was added or is not required

### CODE REVIEW

File: go/vt/vtgate/planbuilder/testdata/aggr_cases.json [hunk 1]
```diff
@@ -6442,5 +6442,119 @@
...
+                          },
+                          "FieldQuery": "select count(*), `user`.foo, `user`.col from `user` where 1 != 1 group by `user`.foo",
+                          "Query": "select count(*), `user`.foo, `user`.col from `user` group by `user`.foo",
+                          "Table": "`user`"
+                        },
+                        {
+                          "OperatorType": "Aggregate",
+                          "Variant": "Ordered",
+                          "Aggregates": "count_star(0), any_value(3)",
+                          "GroupBy": "(1|2)",
```
Comment (GuptaManan100) [inline go/vt/vtgate/planbuilder/testdata/aggr_cases.json:6506]: I don't think this aggregation is correct. Grouping by `ue.bar` and selecting any value of `ue.col` before the left join feels incorrect. The order of operations should be the left join first, and then the grouping and selecting any value.

Comment (systay) [reply go/vt/vtgate/planbuilder/testdata/aggr_cases.json]: found it...

**Review by GuptaManan100 — APPROVED** [review]

File: go/vt/vtgate/planbuilder/operators/apply_join.go [hunk 2]
```diff
@@ -101,8 +103,8 @@ func (aj *ApplyJoin) Clone(inputs []Operator) Operator {
 	kopy.LHS = inputs[0]
 	kopy.RHS = inputs[1]
 	kopy.Columns = slices.Clone(aj.Columns)
-	kopy.JoinColumns = slices.Clone(aj.JoinColumns)
-	kopy.JoinPredicates = slices.Clone(aj.JoinPredicates)
+	kopy.JoinColumns = &applyJoinColumns{columns: slices.Clone(aj.JoinColumns.columns)}
+	kopy.JoinPredicates = &applyJoinColumns{columns: slices.Clone(aj.JoinPredicates.columns)}
```
Comment (frouioui) [inline go/vt/vtgate/planbuilder/operators/apply_join.go:107]: nit: perhaps we should have a `Clone` method on `applyJoinColumns`

File: go/vt/vtgate/planbuilder/operators/hash_join.go [hunk 3]
```diff
@@ -69,14 +82,15 @@ func NewHashJoin(lhs, rhs Operator, outerJoin bool) *HashJoin {
...
 }
 
 func (hj *HashJoin) Clone(inputs []Operator) Operator {
 	kopy := *hj
 	kopy.LHS, kopy.RHS = inputs[0], inputs[1]
-	kopy.columns = slices.Clone(hj.columns)
+	kopy.columns = &hashJoinColumns{columns: slices.Clone(hj.columns.columns)}
 	kopy.LHSKeys = slices.Clone(hj.LHSKeys)
 	kopy.RHSKeys = slices.Clone(hj.RHSKeys)
 	return &kopy
```
Comment (frouioui) [inline go/vt/vtgate/planbuilder/operators/hash_join.go:96]: We don't seem to be cloning the `HashJoin.JoinComparisons` field, is that okay?

**Review by frouioui — APPROVED** [review]

File: go/vt/vtgate/planbuilder/operators/hash_join.go (hunk 3 above)
Comment (systay) [reply go/vt/vtgate/planbuilder/operators/hash_join.go:96]: no it is not! good catch

**Pull request merged**